src/cursor/cur_std.c

```diff
@@ -556,6 +581,215 @@ err:		cursor->saved_err = ret;
 	API_END(session, ret);
 }
 
+/*
+ * __wt_cursor_cache --
+ *	Add this cursor to the cache.
+ */
+int
+__wt_cursor_cache(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
+{
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_ASSERT(session, !F_ISSET(cursor, WT_CURSTD_CACHED) &&
+	    dhandle != NULL);
+
+	WT_TRET(cursor->reset(cursor));
+
+	/*
+	 * Acquire a reference while decrementing the in-use counter.
+	 * After this point, the dhandle may be marked dead, but the
+	 * actual handle won't be removed.
+	 */
+	session->dhandle = dhandle;
+	WT_DHANDLE_ACQUIRE(dhandle);
+	__wt_cursor_dhandle_decr_use(session);
+
+	/* Move the cursor from the open list to the caching hash table. */
+	if (cursor->cache_bucket < 0)
+		cursor->cache_bucket = __wt_hash_city64(
+		    cursor->uri, strlen(cursor->uri)) % WT_HASH_ARRAY_SIZE;
+	TAILQ_REMOVE(&session->cursors, cursor, q);
+	TAILQ_INSERT_HEAD(
+	    &session->cursor_cache[cursor->cache_bucket], cursor, q);
+
+	(void)__wt_atomic_sub32(&S2C(session)->open_cursor_count, 1);
+	WT_STAT_DATA_DECR(session, session_cursor_open);
+	WT_STAT_DATA_INCR(session, session_cursor_cached);
+	F_SET(cursor, WT_CURSTD_CACHED);
+	return (ret);
+}
+
+/*
+ * __wt_cursor_reopen --
+ *	Reopen this cursor from the cached state.
+ */
+void
+__wt_cursor_reopen(WT_CURSOR *cursor, WT_DATA_HANDLE *dhandle)
+{
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cursor->session;
+	WT_ASSERT(session, F_ISSET(cursor, WT_CURSTD_CACHED));
+
+	if (dhandle != NULL) {
+		session->dhandle = dhandle;
+		__wt_cursor_dhandle_incr_use(session);
+		WT_DHANDLE_RELEASE(dhandle);
+	}
+	(void)__wt_atomic_add32(&S2C(session)->open_cursor_count, 1);
+	WT_STAT_DATA_INCR(session, session_cursor_open);
+	WT_STAT_DATA_DECR(session, session_cursor_cached);
+
+	TAILQ_REMOVE(&session->cursor_cache[cursor->cache_bucket], cursor, q);
+	TAILQ_INSERT_HEAD(&session->cursors, cursor, q);
+	F_CLR(cursor, WT_CURSTD_CACHED);
+}
+
+/*
+ * __wt_cursor_cache_release --
+ *	Put the cursor into a cached state, called during close operations.
```

Comment:
Minor nit, add "called during cursor close operations". Not closing the session or connection.

Reply:
Thanks, fixed.